pkg/agent/v1alpha1/server.go

```diff
@@ -0,0 +1,134 @@
+/*
+Copyright 2017 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package utils
+
+import (
+	"net"
+	"os"
+	"sync"
+
+	"github.com/golang/glog"
+	"google.golang.org/grpc"
+
+	"github.com/container-storage-interface/spec/lib/go/csi"
+)
+
+// TD: @amitd
+// This entire code needs to be structured & designed.
+// It is far far away from SOLID principles !!!!
+
+// NonBlockingGRPCServer defines Non blocking GRPC server interfaces
+type NonBlockingGRPCServer interface {
+	// Start services at the endpoint
+	Start(endpoint string, ids csi.IdentityServer, cs csi.ControllerServer, ns csi.NodeServer)
+
+	// Waits for the service to stop
+	Wait()
+
+	// Stops the service gracefully
+	Stop()
+
+	// Stops the service forcefully
+	ForceStop()
+}
+
+// NewNonBlockingGRPCServer returns a new instance of NonBlockingGRPCServer
+func NewNonBlockingGRPCServer() NonBlockingGRPCServer {
+	return &nonBlockingGRPCServer{}
+}
+
+// NonBlocking server
```

Comment:
Need to explain a bit for contributors to refactor/enhance later.

Reply:
done